internal/webhooks/patch_validation.go

```diff
@@ -0,0 +1,365 @@
+						fmt.Sprintf("variable with name %s cannot be found", *jsonPatch.ValueFrom.Variable),
+					))
+			}
+		}
+	}
+	return allErrs
+}
+
+// This contains a list of all of the valid builtin variables.
+// TODO(killianmuldoon): Match this list to controllers/topology/internal/extensions/patches/variables as those structs become available across the code base i.e. public or top-level internal.
+var builtinVariables = sets.NewString(
+	"builtin",
+
+	// Cluster builtins.
+	"builtin.cluster",
+	"builtin.cluster.name",
+	"builtin.cluster.namespace",
+
+	// ClusterTopology builtins.
+	"builtin.cluster.topology",
+	"builtin.cluster.topology.version",
+	"builtin.cluster.topology.class",
+
+	// ControlPlane builtins.
+	"builtin.controlPlane",
+	"builtin.controlPlane.version",
+	"builtin.controlPlane.replicas",
+
+	// MachineDeployment builtins.
+	"builtin.machineDeployment",
+	"builtin.machineDeployment.version",
+	"builtin.machineDeployment.class",
+	"builtin.machineDeployment.name",
+	"builtin.machineDeployment.topologyName",
+	"builtin.machineDeployment.replicas",
+)
+
+func getClusterClassVariablesMap(clusterClassVariables []clusterv1.ClusterClassVariable) map[string]*clusterv1.ClusterClassVariable {
+	variablesMap := map[string]*clusterv1.ClusterClassVariable{}
+	for i := range clusterClassVariables {
+		variablesMap[clusterClassVariables[i].Name] = &clusterClassVariables[i]
+	}
+	return variablesMap
+}
+
+// validateIndexAccess checks to see if the jsonPath is attempting to add an element in the array i.e. access by number
+// If the operation is add an error is thrown if a number greater than 0 is used as an index.
+// If the operation is replace an error is thrown if an index is used.
+func validateIndexAccess(jsonPatch clusterv1.JSONPatch, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	pathParts := strings.Split(jsonPatch.Path, "/")
+	for _, part := range pathParts {
+		// Check if the path segment is a valid number. If an error is thrown continue to the next segment.
+		index, err := strconv.Atoi(part)
+		if err != nil {
+			continue
+		}
+
+		// If the operation is add an error is thrown if a number greater than 0 is used as an index.
+		if jsonPatch.Op == "add" && index != 0 {
+			allErrs = append(allErrs,
+				field.Invalid(path,
+					jsonPatch.Path,
+					"arrays can only be accessed using \"0\" (prepend) or \"-\" (append)",
+				))
+		}
+
+		// If the jsonPatch operation is replace or remove disallow any number as an element in the path.
+		if jsonPatch.Op == "replace" || jsonPatch.Op == "remove" {
+			allErrs = append(allErrs,
+				field.Invalid(path,
+					jsonPatch.Path,
+					fmt.Sprintf("elements in arrays can not be accessed in a %s operation", jsonPatch.Op),
+				))
+		}
+	}
+	return allErrs
+}
+
+func prettyPrint(v interface{}) string {
+	b, err := json.MarshalIndent(v, "", "  ")
+	if err != nil {
+		return errors.Wrapf(err, "failed to marshal value for printing").Error()
```

Comment:
I just want to make sure the source of the error is clear - "failed to marshal" is a really common error across the code base - do you have a suggestion of a way to make the error unique instead of using 'printing'?

Reply:
Ah got it. 

I think we can change it to `failed to marshal field value` which is very unique to the current situation.

Apart from that I think the source is kind of clear because those errors are only used in the context of a field error which will add way more context too.